homeassistant/components/bluetooth/passive_update_coordinator.py

```diff
@@ -0,0 +1,304 @@
+            ),
+        ]
+
+        @callback
+        def _async_cancel_all() -> None:
+            """Cancel all the callbacks."""
+            for cancel in cancels:
+                cancel()
+
+        return _async_cancel_all
+
+    @callback
+    def async_add_entities_listener(
+        self,
+        entity_class: type[PassiveBluetoothCoordinatorEntity],
+        async_add_entites: AddEntitiesCallback,
+    ) -> Callable[[], None]:
+        """Add a listener for new entities."""
+        created: set[PassiveBluetoothEntityKey] = set()
+
+        @callback
+        def _async_add_or_update_entities(
+            data: PassiveBluetoothDataUpdate[_T] | None,
+        ) -> None:
+            """Listen for new entities."""
+            if data is None:
+                return
+            entities: list[PassiveBluetoothCoordinatorEntity] = []
+            for entity_key, description in data.entity_descriptions.items():
+                if entity_key not in created:
+                    entities.append(entity_class(self, entity_key, description))
+                    created.add(entity_key)
+            if entities:
+                async_add_entites(entities)
+
+        return self.async_add_listener(_async_add_or_update_entities)
+
+    @callback
+    def async_add_listener(
+        self,
+        update_callback: Callable[[PassiveBluetoothDataUpdate[_T] | None], None],
+    ) -> Callable[[], None]:
+        """Listen for all updates."""
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self._listeners.remove(update_callback)
+
+        self._listeners.append(update_callback)
+        return remove_listener
+
+    @callback
+    def async_add_entity_key_listener(
+        self,
+        update_callback: Callable[[PassiveBluetoothDataUpdate[_T] | None], None],
+        entity_key: PassiveBluetoothEntityKey | None = None,
+    ) -> Callable[[], None]:
+        """Listen for updates by device key."""
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self._entity_key_listeners[entity_key].remove(update_callback)
+            if not self._entity_key_listeners[entity_key]:
+                del self._entity_key_listeners[entity_key]
+
+        self._entity_key_listeners.setdefault(entity_key, []).append(update_callback)
+        return remove_listener
+
+    @callback
+    def async_update_listeners(
+        self, data: PassiveBluetoothDataUpdate[_T] | None
+    ) -> None:
+        """Update all registered listeners."""
+        # Dispatch to listeners without a filter key
+        for update_callback in self._listeners:
+            update_callback(data)
+
+        # Dispatch to listeners with a filter key
+        for key in self._entity_key_listeners:
+            if listeners := self._entity_key_listeners.get(key):
+                for update_callback in listeners:
+                    update_callback(data)
```

Comment:
Can this be rewritten like this?
```suggestion
        for listeners in self._entity_key_listeners.values():
            if listeners:
                for update_callback in listeners:
                    update_callback(data)
```

Reply:
Although I don't see any filtering happening so maybe the code is wrong

Reply:
The value is always a list, and if empty the key is removed. So I think we can also drop `if listeners`.